solr/core/src/java/org/apache/solr/storage/CompressingDirectory.java

```diff
@@ -0,0 +1,446 @@
+    private int prevBlockSize = BLOCK_SIZE_ESTIMATE; // estimate 50% compression
+    private boolean wroteBlock = false;
+    private final ByteBuffer initialBlock;
+
+    private long filePos;
+    private boolean isOpen;
+
+    /**
+     * Creates a new instance of DirectIOIndexOutput for writing index output with direct IO
+     * bypassing OS buffer
+     *
+     * @throws UnsupportedOperationException if the JDK does not support Direct I/O
+     * @throws IOException if the operating system or filesystem does not support support Direct I/O
+     *     or a sufficient equivalent.
+     */
+    public DirectIOIndexOutput(
+        Path path,
+        String name,
+        int blockSize,
+        int bufferSize,
+        ExecutorService ioExec,
+        int expectLength,
+        boolean useAsyncIO,
+        boolean useDirectIO)
+        throws IOException {
+      super("DirectIOIndexOutput(path=\"" + path.toString() + "\")", name);
+
+      // stored only to lazily compute the pathHash
+      writeHelper = new AsyncDirectWriteHelper(blockSize, bufferSize, path, useDirectIO);
+      buffer = writeHelper.init(0);
+      preBuffer = ByteBuffer.wrap(compressBuffer);
+      initialBlock = ByteBuffer.allocateDirect(blockSize + blockSize - 1).alignedSlice(blockSize);
+      buffer.putLong(0);
+      buffer.putInt(0);
+
+      if (expectLength > bufferSize && useAsyncIO) {
+        writeHelper.start(ioExec);
+      } else {
+        writeHelper.startSync();
+      }
+      isOpen = true;
+    }
+
+    @Override
+    public void writeByte(byte b) throws IOException {
+      preBuffer.put(b);
+      if (!preBuffer.hasRemaining()) {
+        dump();
+      }
+    }
+
+    @Override
+    public void writeBytes(byte[] src, int offset, int len) throws IOException {
+      int toWrite = len;
+      while (true) {
+        final int left = preBuffer.remaining();
+        if (left <= toWrite) {
+          preBuffer.put(src, offset, left);
+          toWrite -= left;
+          offset += left;
+          dump();
+        } else {
+          preBuffer.put(src, offset, toWrite);
+          break;
+        }
+      }
+    }
+
+    private void dump() throws IOException {
+      assert preBuffer.position() == COMPRESSION_BLOCK_SIZE;
+
+      preBuffer.rewind();
+
+      LZ4.compressWithDictionary(compressBuffer, 0, 0, COMPRESSION_BLOCK_SIZE, out, ht);
+      int nextBlockSize = out.resetSize();
+      blockDeltas.writeZInt(nextBlockSize - prevBlockSize);
+      prevBlockSize = nextBlockSize;
+      filePos += COMPRESSION_BLOCK_SIZE;
+
+      preBuffer.clear();
+    }
+
+    private void flush() throws IOException {
+      preBuffer.flip();
+      int preBufferRemaining = preBuffer.remaining();
+      if (preBufferRemaining > 0) {
+        filePos += preBufferRemaining;
+        LZ4.compressWithDictionary(compressBuffer, 0, 0, preBufferRemaining, out, ht);
+      }
+      int blockMapFooterSize = blockDeltas.transferTo(out);
+      if (wroteBlock) {
```

Comment:
Can we have a separate methods for header/footer? 

1. What we are writing on those
2. Also, good to ad byte compression type (enum LZ4=1, gzip=2)

Reply:
They are really written together, so IMO it wouldn't make sense to separate them -- the header is the last thing written. The footer is written immediately for, by `int blockMapFooterSize = blockDeltas.transferTo(out)`, and then two things are written to the space pre-allocated for the header: the total logical length of the file, and the size of the serialized footer. The footer itself contains delta-encoded vint sizes of the individual compressed blocks, to allow mapping logical file blocks to their associated offset/length in the compressed file

Reply:
1. See if we can add blocksize as well. Also any other meta data
2. In general, function names are self documentation. Would be good to add documentation for that class, format, etc.

Reply:
a2ad9cf9f40ff673c7436aae2ba6c74bb2eee210 added compression type and block size (2 bytes), plus added 2 extra bytes for potential further extension without reindexing. The implementation barely changed (all static final fields, assuming the same type of compression), but the on-disk format records the metadata, the write code writes it, and the read code verifies it, so we can extend this to respect these values in a dynamic way later if/when need be.